File: 2021/day-02/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

func partOne(scanner *bufio.Scanner) {
	var depth, horizontal int

	for scanner.Scan() {
		instruction := strings.Split(scanner.Text(), " ")

		distance, err := strconv.Atoi(instruction[1])
		if err != nil {
			log.Fatal(err)
		}

		switch instruction[0] {
		case "up":
			depth -= distance
		case "down":
			depth += distance
		case "forward":
			horizontal += distance
		}
	}

	fmt.Println("Depth:", depth, "Distance:", horizontal, "Product:", (depth * horizontal))
}

func partTwo(scanner *bufio.Scanner) {
	var depth, horizontal, aim int

	for scanner.Scan() {
		instruction := strings.Split(scanner.Text(), " ")

		distance, err := strconv.Atoi(instruction[1])
		if err != nil {
			log.Fatal(err)
		}

		switch instruction[0] {
		case "up":
			aim -= distance
		case "down":
			aim += distance
		case "forward":
			horizontal += distance
			depth += (aim * distance)
		}
	}

	fmt.Println("Depth:", depth, "Distance:", horizontal, "Product:", (depth * horizontal))
}

func main() {
	buffer := 1
	var err error
	if len(os.Args) > 1 {
		if buffer, err = strconv.Atoi(os.Args[1]); err != nil {
			log.Fatal(("Could not convert arg to number: " + os.Args[1]))
		}
	}
	switch buffer {
	case 1:
		partOne(bufio.NewScanner(os.Stdin))
	case 2:
		partTwo(bufio.NewScanner(os.Stdin))
	}
}

File: 2021/day-02/main.cc
#include <iostream>
#include <string>

void part1() {
  int depth = 0;
  int horizontal = 0;

  std::string line;
  while (std::cin >> line) {
    std::string direction = line;

    std::cin >> line;
    int distance = std::stoi(line);

    if (direction == "up") {
      depth = depth - distance;
    } else if (direction == "down") {
      depth = depth + distance;
    } else if (direction == "forward") {
      horizontal = horizontal + distance;
    }
  }

  std::cout << "Depth: " << depth << "\tDistance: " << horizontal
            << "\tProduct: " << (depth * horizontal) << std::endl;
}

void part2() {
  int depth = 0;
  int horizontal = 0;
  int aim = 0;

  std::string line;
  while (std::cin >> line) {
    std::string direction = line;

    std::cin >> line;
    int distance = std::stoi(line);

    if (direction == "up") {
      aim -= distance;
    } else if (direction == "down") {
      aim += distance;
    } else if (direction == "forward") {
      horizontal += distance;
      depth += (aim * distance);
    }
  }

  std::cout << "Depth: " << depth << "\tDistance: " << horizontal
            << "\tProduct: " << (depth * horizontal) << std::endl;
}

int main(int argc, char** argv) {
  int buffer = 1;
  if (argc > 1) {
    buffer = std::atoi(argv[1]);
  }

  if (buffer == 1) {
    part1();
  } else if (buffer == 2) {
    part2();
  } else {
    std::cout << "No valid part number selected." << std::endl;
  }

  return 0;
}
